src/operator/sum.rs

```diff
@@ -0,0 +1,201 @@
+//! N-ary plus operator.
+
+use crate::{
+    algebra::{AddAssignByRef, AddByRef, HasZero, WithSize},
+    circuit::{
+        operator_traits::{NaryOperator, Operator},
+        Circuit, OwnershipPreference, Scope, Stream,
+    },
+};
+use std::{borrow::Cow, iter::once, mem::transmute, ops::Add};
+
+impl<P, D> Stream<Circuit<P>, D>
+where
+    P: Clone + 'static,
+    D: Add<Output = D> + AddByRef + AddAssignByRef + Clone + HasZero + WithSize + 'static,
+{
+    /// Apply the [`Sum`] operator to `self` and all streams in `streams`.
+    pub fn sum<'a, I>(&'a self, streams: I) -> Stream<Circuit<P>, D>
+    where
+        I: Iterator<Item = &'a Self>,
+    {
+        self.circuit()
+            .add_nary_operator(Sum::new(), once(self).chain(streams))
+    }
+}
+
+/// Operator that computes the sum of values across all its input streams at
+/// each timestamp.
+pub struct Sum<D>
+where
+    D: Clone + 'static,
+{
+    // Vector of input values.  Keep it here to reuse the allocation across
+    // operator invocations.
+    inputs: Vec<Cow<'static, D>>,
+}
+
+impl<D> Sum<D>
+where
+    D: Clone + 'static,
+{
+    pub fn new() -> Self {
+        Self { inputs: Vec::new() }
+    }
+}
+
+impl<D> Default for Sum<D>
+where
+    D: Clone + 'static,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<D> Operator for Sum<D>
+where
+    D: Clone + 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Sum")
+    }
+
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+impl<D> NaryOperator<D, D> for Sum<D>
+where
+    D: Add<Output = D> + AddAssignByRef + Clone + HasZero + WithSize + 'static,
+{
+    fn eval<'a, Iter>(&mut self, inputs: Iter) -> D
+    where
+        Iter: Iterator<Item = Cow<'a, D>>,
+    {
+        let input_vec = unsafe {
+            transmute::<&mut Vec<Cow<'static, D>>, &mut Vec<Cow<'a, D>>>(&mut self.inputs)
+        };
+        input_vec.clear();
+        for input in inputs {
+            input_vec.push(input);
+        }
+        input_vec.sort_by(|x, y| y.size().cmp(&x.size()));
+
+        if input_vec.is_empty() {
+            return D::zero();
+        }
+        let mut res = D::zero();
+        for input in input_vec.drain(..) {
```

Comment:
So I guess this saves n-1 intermediate results and scheduler invocations.
But on the other hand you sort this vector every time. It's true it has size 3 so far.

Reply:
This doesn't reduce the number of intermediate results, but hopefully produces them in a more optimal order due to the sorting...